File: 2024/ocaml/bin/main.ml
open In_channel

let () =
  let args = Sys.argv in
  if Array.length args != 3 then
    Printf.printf "Expected 2 arguments, got %d\n" (Array.length args)
  else
    let day, part =
      (Array.get args 1, String.lowercase_ascii (Array.get args 2))
    in
    let input = input_all (open_in ("inputs/" ^ day)) in
    let out =
      match (day, part) with
      | "4", "a" -> Solutions.Day4.solve_a input
      | "4", "b" -> Solutions.Day4.solve_b input
      | _ -> Printf.sprintf "Day %s part %s was not solved in OCaml\n" day part
    in
    print_endline ("Solution was " ^ out)

File: 2024/cpp/main.cpp
#include "day8.h"
#include <fstream>
#include <print>
#include <sstream>
#include <unordered_map>

using Solver = std::string (*)(std::string_view);

using namespace std::string_view_literals;

std::unordered_map<std::string_view, std::unordered_map<std::string_view, Solver>> solutions {
    { "8", { { "a", Day8::solve_a }, { "b", Day8::solve_b } } },

};

std::string read_to_string(std::string_view path)
{
    std::ifstream in { path.data() };
    std::ostringstream os;
    os << in.rdbuf();
    return os.str();
}

int main(int argc, char* argv[])
{
    if (argc != 3) {
        std::println("Usage: AdventOfCode <day> <part>");
        return 1;
    }

    auto day = argv[1];
    auto part = argv[2];

    auto input = read_to_string(std::format("inputs/{}", day));

    std::println("The solution was: {}", solutions.at(day).at(part)(input));

    return 0;
}
